### PR TITLE
Feature/issue 1957 cov exp quad

### PR DESCRIPTION
**Pull request by rtrangucci**
#### TODO

- [x] Add doc to the language.

#### Submisison Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

Expose cov_exp_quad and squared_distance(real, real) to language.

#### Intended Effect

Define allowable function signatures and good and bad model unit tests.

#### How to Verify

bad models: `/runTests.py src/test/unit/lang/parser/other_test.cpp `
good models: `./runTests.py src/test/unit/lang/parser/matrix_functions_test.cpp`
header-tests: 
- `make src/test/test-models/good/function-signatures/math/matrix/cov_exp_quad.hpp-test`
- `make src/test/test-models/good/function-signatures/math/matrix/squared_distance.hpp-test`

#### Side Effects

None

#### Documentation

Need to update doc for cov_exp_quad and Gaussian Process chapter in manual.

#### Reviewer Suggestions

@syclik 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Rob Trangucci

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/test/test-models/good/function-signatures/math/matrix/cov_exp_quad.stan [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+  transformed_param_matrix <- cov_exp_quad(p_arr_1, d_arr_2, p_sigma, p_len);
+  transformed_param_matrix <- cov_exp_quad(p_arr_1, d_arr_2, d_sigma, p_len);
+  transformed_param_matrix <- cov_exp_quad(p_arr_1, d_arr_2, p_sigma, d_len);
+  transformed_param_matrix <- cov_exp_quad(p_arr_1, d_arr_2, d_sigma, d_len);
+
+  transformed_param_matrix <- cov_exp_quad(d_arr_1, d_arr_2, p_sigma, p_len);
+  transformed_param_matrix <- cov_exp_quad(d_arr_1, d_arr_2, d_sigma, p_len);
+  transformed_param_matrix <- cov_exp_quad(d_arr_1, d_arr_2, p_sigma, d_len);
+
```
Comment (syclik) [inline src/test/test-models/good/function-signatures/math/matrix/cov_exp_quad.stan]: missing `transformed_param_matrix <- cov_exp_quad(d_arr_1, d_arr_2, d_sigma, d_len);`

**Comment by syclik** [comment]
@rtrangucci: we're missing doc for the manual. We need to include it.

**Comment by bob-carpenter** [comment]
I can help with the doc if there are definitions somewhere of what the function does other than in the code.  Also, we've just been putting the model tests for functions in a single file rather than having them this split out.  We should decide which way to go (I think we should keep them together in a single file, because it's easier that way to see if all the permutations are instantiated).

**Comment by rtrangucci** [comment]
@bob-carpenter I'm going to add doc for this today. What tex files do I need to update in order to properly doc everything? `functions.tex` and the chapter on Gaussian Processes in `programming.tex`?

**Comment by bob-carpenter** [comment]
Just functions.tex to doc the new functions.

But if you want people to find it, the programming.tex
examples.

Or even better, write a case study.  I'm moving to doing
that instead of just writing more manual examples since everyone
said the manual was getting too large and wasn't detailed
enough about practical matters in fitting models.

- Bob

**Comment by rtrangucci** [comment]
@bob-carpenter a case study sounds like a good idea.

How does one make the manual? I'm getting a font error here:

```
kpathsea: Running mktexpk --mfmode / --bdpi 600 --mag 0+420/600 --dpi 420 hlsrot8r
mktexpk: don't know how to create bitmap font for hlsrot8r.
mktexpk: perhaps hlsrot8r is missing from the map file.
kpathsea: Appending font creation commands to missfont.log.
 )
!pdfTeX error: pdflatex (file hlsrot8r): Font hlsrot8r at 420 not found
 ==> Fatal error occurred, no output PDF file produced!
Latexmk: Index file 'stan-reference.idx' was written
Latexmk: Found input bbl file 'stan-reference.bbl'
Latexmk: Found bibliography file(s) [../bibtex/all.bib]
Latexmk: Errors, so I did not complete making targets
Collected error summary (may duplicate other messages):
  pdflatex: Command for 'pdflatex' gave return code 256
Latexmk: Use the -f option to force complete processing,
 unless error was exceeding maximum runs of latex/pdflatex.
make: *** [src/docs/stan-reference/stan-reference.pdf] Error 12
```

**Comment by bob-carpenter** [comment]
No idea what's going on there or what mktexpk does.

I'm using the Lucida fonts, but they're not free and
they're a pain in the ass to install.

What kind of computer and which version of LaTeX are
you using?  LaTeX/TeX is such a dinosaur that I find it
almost impossible to debug.  

- Bob

**Comment by rtrangucci** [comment]
@bob-carpenter this is a problem specific to my latex install, so I'm just going to build the manual without Lucida fonts. I'll try to reinstall `lucimatx` at some later point.

**Comment by syclik** [comment]
Looks good! I gave @rtrangucci some doc comments in person. Once they're in, we can merge.

**Comment by bob-carpenter** [comment]
It should check to see if they're installed and
revert to built-ins otherwise (it won't be guaranteed
to not have overfull hboxes, but it should build).

**Comment by rtrangucci** [comment]
This is probably ready to merge.

**Pull request merged**